olive/engine/cloud_cache_helper.py

```diff
@@ -0,0 +1,244 @@
+        self.container_client = ContainerClient(
+            account_url=account_url, container_name=self.container_name, credential=credential
+        )
+
+        self.map_blob = "model_map.json"
+        self.local_map_path = Path(cache_dir) / "model_map.json"
+        self.local_map_path.parent.mkdir(parents=True, exist_ok=True)
+        self.cache_dir = cache_dir
+        self.output_model_path = Path(cache_dir) / "model"
+        self.cloud_cache_map = self.download_model_cache_map()
+        self.input_model_config = input_model_config
+        self.hf_cache_path = hf_cache_path
+
+    def update_model_config(self, cloud_model_path, model_config, input_model_hash):
+        logger.info("Updating model config with cloud model path: %s", cloud_model_path)
+        model_directory_prefix = f"{input_model_hash}/model"
+        blob_list = self.container_client.list_blobs(name_starts_with=model_directory_prefix)
+
+        for blob in blob_list:
+            blob_client = self.container_client.get_blob_client(blob)
+            local_file_path = self.output_model_path / blob.name[len(model_directory_prefix) + 1 :]
+            logger.info("Downloading %s to %s", blob.name, local_file_path)
+            local_file_path.parent.mkdir(parents=True, exist_ok=True)
+            with open(local_file_path, "wb") as download_file:
+                download_stream = blob_client.download_blob()
+                download_file.write(download_stream.readall())
+
+        model_config.config["model_path"] = str(self.output_model_path / cloud_model_path)
+
+        return model_config
+
+    def download_model_cache_map(self):
+        logger.info("Downloading model cache map from %s", self.map_blob)
+        blob_client = self.container_client.get_blob_client(self.map_blob)
+
+        if not blob_client.exists():
+            return {}
+
+        self.local_map_path.unlink(missing_ok=True)
+        self.local_map_path.touch()
+
+        with open(self.local_map_path, "wb") as download_file:
+            download_stream = blob_client.download_blob()
+            download_file.write(download_stream.readall())
+
+        with open(self.local_map_path) as file:
+            return json.load(file)
+
+    def get_hash_key(self, model_config: ModelConfig, pass_search_point: Dict[str, Any], input_model_hash: str):
+        hf_hub_model_commit_id = None
+        if input_model_hash is None:
+            model_config_copy = self.input_model_config
+            model_path = model_config_copy.config.get("model_path")
+            if model_path is None and model_config_copy.config.get("hf_config") is not None:
+                # hf model, get model commit from hf cache
+                cache_dir = Path(self.hf_cache_path)
+                model_cache_dir = Path(cache_dir) / "hub"
+                model_name = f"models--{model_config_copy.config['hf_config']['model_name'].replace('/', '--')}"
+                commit_path = model_cache_dir / model_name / "refs" / "main"
```

Comment:
Seems like this logic basing decisions on undocumented internals of how hfhub works. Is there no documented API to get some of this information? This can likely change tomorrow behind our back and break entire caching logic. On the other end, is this path even required to be part of the hash? There should be an explicit _commit\_id_ or something that hfhub uses.

Reply:
There is no documentation about this. and yes this might be fragile but I don't think this is a normal use case and they won't have an API for this. In case it breaks the logic, I can add a fall-back logic here to skip caching if it breaks.

You are right this path is the commit-id they use. This `main` file just includes a commit id.

Reply:
https://github.com/huggingface/huggingface_hub/issues/2045#issuecomment-1963787595
```
>>> from huggingface_hub import model_info

>>> model_info("google/gemma-7b-it").sha
'bee54ce1bd2a5b36d68a06857fc0871a3389ca0b'

>>> model_info("google/gemma-7b-it", revision="flax").sha
'90fe4a773beffece715c57300f9f2c31b230724b'

>>> model_info("google/gemma-7b-it", revision="float16").sha
'cda5a5189d4606b969ca2b316298debf5fc2953d'
```

Reply:
Or even better, https://github.com/huggingface/huggingface_hub/issues/2045#issuecomment-1964137527

Reply:
Good to know this! Not sure if the local huggingface api will always use the latest commit from hub. Will it check the commit id difference between local and hub. But I assume this should be always true if the we don't specify the commit info or version info.

Reply:
Seems this is not true. I have tested to load `Intel/bert-base-uncased-mrpc` model (from my local cache), however, the commit id is not updated to the latest one: https://huggingface.co/Intel/bert-base-uncased-mrpc/commit/844e804944328e45c3f3bf69b35467e0ce0b26af. Can you double check this?

Reply:
Not sure how Huggingface cache works, but if you just directly load model from pretrained api in Python (without using Olive), seems it will not fetch the latest commit. But somehow if i run Olive, it can fetch the latest commit from hf hub.

Reply:
Yes, look like the behavior is inconsistent between APIs. Olive uses `AutoModel.from_pretrained` which passes `revision=None` as input to other APIs but `PretrainedModel.from_pretrained` passes `revision=main`. This difference is the source of the discrepancy. I don't think it is still wrong as long as the usage works consistently within Olive. If the commit hash that the API returns is the same as the model that is loaded, we are good. Whether the _newest_ model is downloaded or not is secondary problem outside the scope of this PR.

Reply:
Yeah the concern here is user loaded model locally commit id is inconsistent with remote API. That's the reason I would prefer to check local cache. Remote API can be inconsistent but local cache is always right.

Reply:
That shouldn't be inconsistent because both models are being loaded from Olive. The inconsistency is only if one end loads in Olive and the other outside Olive.

I still think using the available API is better than hardcoding some undocumented paths. I would fallback to this solution only if we have a demonstrable scenario where both ends mismatch. Alternatively, we could report this as a bug on HuggingFace repo.

Reply:
Yeah i have already made a change here to use HF api. The inconsistence comes from, when Olive loads model from local cache with old commit id, but we always call the remote API to fetch the latest commit id for hashing, turns out the hash key is wrong.

Reply:
Since as for my testing, seems Olive somehow will download and use the latest commit from HF, I'm good for this API now. But I still have concerns this could be potential issue and this would be hard to debug in the future if there was any issue.